index.ts

```diff
@@ -1,35 +1,60 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private _keys: string[] | undefined
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(data?: object[], options?: { keys?: string[] }) {
+    if (data !== undefined) { this.push(...data) }
+    if (options !== undefined) {
+      if (options.keys !== undefined) { this._keys = options.keys }
+    }
+  }
+
+  get keys(): string[] | undefined {
+    return this._keys
+  }
 
   get length(): number {
     return this._original.length
   }
 
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  get original(): object[] {
+    return this._original
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  get transposed(): object[] {
+    return this._original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this._transposed[keyName] === undefined) {
+      return undefined
+    }
     return this._transposed[keyName][keyValue]
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public push(...data: object[]) {
+    data.forEach((datum) => {
+      const clone = Object.assign({}, datum)
+      this._original.push(clone)
+      this.transpose(clone)
+    })
+    return this._original.length
   }
-}
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  private transpose(datum: { [key: string]: any }) {

```

Comment:
I don't understand the rewrite here and prefer the original; it's more concise and imo more readable. Think `reduce` might also be higher performance than `forEach` used in this rewrite

Reply:
reduce is nowhere near as readable; I still have no clue how it work. I don't see how what is essentially is a nested for, could have better performance than just one

Reply:
Outdated